File: lib_execute_script/src/main/java/org/cmsfs/execute/script/ScriptExecute.java
package org.cmsfs.execute.script;

import org.apache.commons.io.FileUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class ScriptExecute {

    final String workspace = "workspace";

    final long workNumber;
    final String workFolder;

    final String currentPath = System.getProperty("user.dir");

    final List<String> files;
    final Map<String, String> env;
    final Optional<String> dataOpt;
    final Optional<String> argsOpt;

    public ScriptExecute(List<String> files, Map<String, String> env, Optional<String> dataOpt, Optional<String> argsOpt) {
        this.workNumber = ThreadLocalRandom.current().nextLong(100000000);
        this.workFolder = currentPath + "/" + workspace + "/" + workNumber;
        this.files = files;
        this.env = env;
        this.dataOpt = dataOpt;
        this.argsOpt = argsOpt;
    }

    public ScriptExecute(List<String> files, Map<String, String> env) {
        this(files, env, Optional.empty(), Optional.empty());
    }

    public ScriptExecute(List<String> files, Optional<String> data, Optional<String> args) {
        this(files, new HashMap<>(), data, args);
    }

    public ScriptExecute(List<String> files) {
        this(files, new HashMap<>(), Optional.empty(), Optional.empty());
    }

    public ScriptExecute(List<String> files, Optional<String> data) {
        this(files, new HashMap<>(), data, Optional.empty());
    }

    void writeFile(String content, String file) {
        try {
            FileUtils.writeStringToFile(new File(file), content, Charset.forName("UTF-8"), false);
        } catch (IOException ioe) {
            System.out.println(ioe.getMessage());
        }
    }

    public String start() throws Exception {

        String mainFile = files.get(0);

        List<String> executor = executorChoice(mainFile);

        List<String> commands = new ArrayList<>();

        executor.forEach(commands::add);

        createWorkDir();

        files.forEach(this::downScript);

        commands.add(mainFile);

        dataOpt.ifPresent(data -> {
            writeFile(data, this.workFolder + "/" + "data.json");
            commands.add("data.json");
        });

        argsOpt.ifPresent(data -> {
            writeFile(data, this.workFolder + "/" + "args.json");
            commands.add("args.json");
        });


        String[] ss = new String[]{"a", "b", "c"};
        ProcessBuilder pb = new ProcessBuilder(commands)
                .directory(new File(this.workFolder));
        Map<String, String> pbEnv = pb.environment();
        env.forEach((k, v) -> pbEnv.put(k, v));

        Process p = pb.start();

        String line;
        List<String> lines = new ArrayList<>();
        BufferedReader r = new BufferedReader(new InputStreamReader(p.getInputStream()));
        while ((line = r.readLine()) != null) {
            lines.add(line);
        }
        r.close();

        try {
            return String.join("\n", lines);
        } finally {
            FileUtils.forceDeleteOnExit(new File(this.workFolder));
        }
    }

    void createWorkDir() throws IOException {
        String workDirPath = this.workspace + "/" + this.workNumber;
        FileUtils.forceMkdir(new File(workDirPath));
    }

    void downScript(String sFile) {
        System.out.println(sFile);
        try {
            File sourceFile = new File(sFile);
            File targetFile = new File(this.workFolder + "/" + sourceFile.getName());
            FileUtils.copyURLToFile(sourceFile.toURI().toURL(), targetFile);
        } catch (IOException ioe) {
            System.out.println(ioe.getMessage());
        }
    }

    List<String> executorChoice(String mainFile) throws Exception {
        String[] p = mainFile.split("\\.");
        String executer = p[p.length - 1];

        switch (executer) {
            case "py":
                return Arrays.asList("python3");
            case "sh":
                return Arrays.asList("sh");
            case "ps1":
                return Arrays.asList("powershell", "-file");
            default:
                throw new Exception("actuator unknown: " + executer);
        }
    }
}

File: lib_execute_script/src/main/scala/org/cmsfs/lib/execute/script/ScriptExecute.scala
package org.cmsfs.lib.execute.script

import java.io.File
import java.nio.charset.Charset
import java.util.concurrent.ThreadLocalRandom

import org.apache.commons.io.FileUtils
import play.api.libs.json.JsValue

object ScriptExecute {

  private def executorChoice(mainFile: String): Seq[String] = {
    val name = mainFile.split("\\.").last
    name match {
      case "py" => Seq("python3")
      case "sh" => Seq("sh")
      case "ps1" => Seq("powershell")
      case _ => throw new Exception(s"actuator unknown: ${name}")
    }
  }

  private def createWorkDirAndDeleteAfterExecute(): (String, () => Unit) = {
    val dirName = s"workspace/${ThreadLocalRandom.current.nextLong(Long.MaxValue).toString}"
    val dir = new File(dirName)
    FileUtils.forceMkdir(dir)

    (dirName, () => FileUtils.deleteDirectory(dir))
  }

  private def writeFile(fileName: String, content: String): Unit =
    FileUtils.writeStringToFile(new File(fileName), content, Charset.forName("UTF-8"), false)

  private def wirteFile(data: String, dirPath: String, fileName: String): Unit =
    writeFile(s"${dirPath}/${fileName}", data)

  private def executeScript(files: Seq[Seq[String]], env: Map[String, String], dataOpt: Option[String], argsOpt: Option[JsValue]): Option[String] = {
    import sys.process._

    val mainFile = files(0).last

    var executor: Seq[String] = executorChoice(mainFile)
    val (workDirName, deleteWorkDirFun) = createWorkDirAndDeleteAfterExecute()

    executor = executor :+ mainFile

    if (dataOpt.isDefined) {
      wirteFile(dataOpt.get, workDirName, "data.json")
      executor = executor :+ "data.json"
    }

    if (argsOpt.isDefined) {
      wirteFile(argsOpt.get.toString(), workDirName, "args.json")
      executor = executor :+ "args.json"
    }

    val result: Option[String] = try {
      Some(Process(executor, new java.io.File(workDirName), env.toSeq: _*).!!)
    } catch {
      case ex: Exception =>
        println(ex.getMessage)
        None
    }
    deleteWorkDirFun()
    result
  }
}
